### DISCUSSION TITLE
zeros remain zero?

### DISCUSSION THREAD

**Discussion by RanchiZhao** [bitsandbytes-foundation/bitsandbytes#1159] [Q&A]
Does the BNB INT8 quantization method ensure that weight values originally at zero remain zero after the quantization process?

**Comment by gyanu2507** [comment]
Yes for the INT8 *weights*. LLM.int8() is vector-wise absmax: `scale = max(|w|) / 127`, then `q = round(w / scale)`. A true zero stays zero because `round(0 / scale) = 0`, and dequant is `q * scale`.

Two caveats:

1. Hidden-state outliers are *not* stored in INT8. Those columns are kept in fp16 (the "mixed-precision decomposition" in the LLM.int8() paper). A weight that is zero in an outlier column is still zero. Activations in that column just skip the INT8 matmul.
2. If you mean "zeros in the quantized tensor stay bitwise zero after a round-trip through CPU/GPU copies," yes. If you mean "sparsity is preserved as a structured sparse format," no. It is still a dense INT8 matrix that happens to contain zeros.

Row-wise scales (what you see in HF checkpoints) do not change the zero case.
